File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        left = 0
        right = 1
        lst = []
        sell = prices[left]
        while right<len(prices):
            if prices[right]>sell:
                sell = prices[right]
            if prices[right]<prices[left]:
                lst.append(sell-prices[left])
                left = right
                sell = prices[left]
                right += 1
            else:
                right+=1
        lst.append(sell - prices[left])
        return max(lst)

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
// T- O(n)
// S- O(1)
class Solution {
    public int maxProfit(int[] prices) {
        int minn = prices[0];
        int maxP = Integer.MIN_VALUE;
        for (int i = 1; i < prices.length; i++) {
            if (prices[i] > minn) {
                maxP = Math.max(maxP, prices[i] - minn);
            } else {
                minn = prices[i];
            }
        }

        return (maxP == Integer.MIN_VALUE) ? 0 : maxP;
    }
}